File: jms-benchmark/src/main/scala/org/fusesource/jmsbenchmark/Scenario.scala
/**
 * Copyright (C) 2009-2011 the original author or authors.
 * See the notice.md file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.fusesource.jmsbenchmark

import java.util.concurrent.atomic._
import java.util.concurrent.TimeUnit._
import scala.collection.mutable.ListBuffer
import java.util.concurrent.{CountDownLatch, BrokenBarrierException, TimeoutException, CyclicBarrier}

case class DataSample(time:Long, produced:Long, consumed:Long, errors:Long, max_latency:Long)

abstract class SleepFn {
  def init(time:Long):Unit
  def apply(client:Scenario#Client):Long
}

object Scenario {
  val MESSAGE_ID:Array[Byte] = "message-id"
  val NEWLINE = '\n'.toByte
  val NANOS_PER_SECOND = NANOSECONDS.convert(1, SECONDS)
  
  implicit def toBytes(value: String):Array[Byte] = value.getBytes("UTF-8")

  def o[T](value:T):Option[T] = value match {
    case null => None
    case x => Some(x)
  }
}

trait Scenario {
  import Scenario._

  var url:String = "tcp://localhost:61616"
  var user_name:String = _
  var password:String = _

  var _producer_sleep: SleepFn = new SleepFn { def apply(client:Scenario#Client) = 0; def init(time: Long) {}  }
  def producer_sleep_= (new_value: Int) = _producer_sleep = new SleepFn{ def apply(client:Scenario#Client) = new_value; def init(time: Long) {}  }

  var _consumer_sleep: SleepFn = new SleepFn { def apply(client:Scenario#Client) = 0; def init(time: Long) {}  }
  def consumer_sleep_= (new_value: Int) = _consumer_sleep = new SleepFn { def apply(client:Scenario#Client) = new_value; def init(time: Long) {}  }

  var producers = 1
  var producers_per_sample = 0

  var consumers = 1
  var consumers_per_sample = 0
  var sample_interval = 1000

  var message_size = 1024
  var persistent = false

  var headers = Array[Array[(String,String)]]()
  var selector:String = null
  var no_local = false
  var durable = false
  var ack_mode = "auto"
  var messages_per_connection = -1L
  var display_errors = false
  var tx_size = 0

  var destination_type = "queue"
  private var _destination_name: () => String = () => "load"
  def destination_name = _destination_name()
  def destination_name_=(new_name: String) = _destination_name = () => new_name
  def destination_name_=(new_func: () => String) = _destination_name = new_func
  var destination_count = 1

  val producer_counter = new AtomicLong()
  val consumer_counter = new AtomicLong()
  val error_counter = new AtomicLong()
  val max_latency = new AtomicLong()
  val done = new AtomicBoolean()

  var queue_prefix = ""
  var topic_prefix = ""
  var name = "custom"

  var drain_timeout = 2000L

  def run() = {
    print(toString)
    println("--------------------------------------")
    println("     Running: Press ENTER to stop")
    println("--------------------------------------")
    println("")

    with_load {

      // start a sampling client...
      val sample_thread = new Thread() {
        override def run() = {
          
          def print_rate(name: String, periodCount:Long, totalCount:Long, nanos: Long) = {

            val rate_per_second: java.lang.Float = ((1.0f * periodCount / nanos) * NANOS_PER_SECOND)
            println("%s total: %,d, rate: %,.3f per second".format(name, totalCount, rate_per_second))
          }

          try {
            var start = System.nanoTime
            var total_producer_count = 0L
            var total_consumer_count = 0L
            var total_error_count = 0L
            collection_start
            while( !done.get ) {
              Thread.sleep(sample_interval)
              val end = System.nanoTime
              collection_sample
              val collected = collection_end

              if ( producers > 0 ) {
                val count = collected.last.produced
                total_producer_count += count
                if ( total_producer_count > 0 ) {
                  print_rate("Producer", count, total_producer_count, end - start)
                }
              }

              if ( consumers > 0 ) {
                val count = collected.last.consumed
                total_consumer_count += count
                if ( total_consumer_count > 0 ) {
                  print_rate("Consumer", count, total_consumer_count, end - start)
                }
              }

              val count = collected.last.errors
              total_error_count += count
              if ( total_error_count > 0 ) {
                if ( count != 0 ) {
                  print_rate("Error", count, total_error_count, end - start)
                }
              }

              start = end
            }
          } catch {
            case e:InterruptedException =>
          }
        }
      }
      sample_thread.start()

      System.in.read()
      done.set(true)

      sample_thread.interrupt
      sample_thread.join
    }

  }

  override def toString() = {
    "--------------------------------------\n"+
    "Scenario Settings\n"+
    "--------------------------------------\n"+
    "  destination_type      = "+destination_type+"\n"+
    "  queue_prefix          = "+queue_prefix+"\n"+
    "  topic_prefix          = "+topic_prefix+"\n"+
    "  destination_count     = "+destination_count+"\n" +
    "  destination_name      = "+destination_name+"\n" +
    "  sample_interval (ms)  = "+sample_interval+"\n" +
    "  \n"+
    "  --- Producer Properties ---\n"+
    "  producers             = "+producers+"\n"+
    "  message_size          = "+message_size+"\n"+
    "  persistent            = "+persistent+"\n"+
    "  producer_sleep (ms)   = "+_producer_sleep(null)+"\n"+
    "  headers               = "+headers.mkString(", ")+"\n"+
    "  \n"+
    "  --- Consumer Properties ---\n"+
    "  consumers             = "+consumers+"\n"+
    "  consumer_sleep (ms)   = "+_consumer_sleep(null)+"\n"+
    "  selector              = "+selector+"\n"+
    "  durable               = "+durable+"\n"+
    ""

  }

  protected def headers_for(i:Int) = {
    if ( headers.isEmpty ) {
      Array[(String, String)]()
    } else {
      headers(i%headers.size)
    }
  }

  var data_samples = ListBuffer[DataSample]()

  def collection_start: Unit = {
    producer_counter.set(0)
    consumer_counter.set(0)
    error_counter.set(0)
    max_latency.set(0)
    data_samples = ListBuffer[DataSample]()
  }

  def update_max_latency(value:Long) = {
    var was = max_latency.get
    while ( value > was && !max_latency.compareAndSet(was, value) ) {
      // someone else changed it before we could update it, lets
      // get the update and see if we still need to update.
      was = max_latency.get
    }
  }

  def collection_end = data_samples.toList

  trait Client {
    def id:Int
    def start():Unit
    def shutdown():Unit
    def wait_for_shutdown:Unit
  }

  var producer_clients = List[Client]()
  var consumer_clients = List[Client]()

  var connected_latch:CountDownLatch = null
  var apply_load_latch:CountDownLatch = null

  def with_load[T](func: =>T ):T = {
    var i = 0
    with_load_and_connect_timeout(func) {
      i += 1
      if( i <= 10 ) {
        1000L
      } else {
        0L
      }
    }
  }

  def with_load_and_connect_timeout[T](func: =>T)(continue_connecting: =>Long):T = {
    done.set(false)

    _producer_sleep.init(System.currentTimeMillis())
    _consumer_sleep.init(System.currentTimeMillis())

    connected_latch = new CountDownLatch(producers + consumers);
    apply_load_latch = new CountDownLatch(1);

    for (i <- 0 until producers) {
      val client = createProducer(i)
      producer_clients ::= client
      client.start()
    }

    for (i <- 0 until consumers) {
      val client = createConsumer(i)
      consumer_clients ::= client
      client.start()
    }

    try {
      var timeout = 0L
      var done_connect = false
      while( !done_connect ) {
        if ( connected_latch.await(timeout, MILLISECONDS) ) {
          done_connect = true
          apply_load_latch.countDown();
        }  else {
          timeout = continue_connecting
          if ( timeout <= 0 ) {
            apply_load_latch.countDown();
            done.set(true)
            throw new TimeoutException("All clients could not connect.  Failing connections: "+connected_latch.getCount);
          }
        }
      }
      func
    } finally {
      done.set(true)
      for( client <- producer_clients ) {
        client.shutdown
      }
      for( client <- producer_clients ) {
        client.wait_for_shutdown
      }
      producer_clients = List()
      // wait for the threads to finish..
      for( client <- consumer_clients ) {
        client.shutdown
      }
      for( client <- consumer_clients ) {
        client.wait_for_shutdown
      }
      consumer_clients = List()
    }
  }

  def drain = {
    done.set(false)
    if( destination_type=="queue" || destination_type=="raw_queue" || durable==true ) {
      print("draining")
      consumer_counter.set(0)
      var drained = 0L

      val original_tx_size = tx_size
      val original_consumer_sleep = _consumer_sleep
      val original_ack_mode = ack_mode

      // Lets change the consumer config to consume as fast as possible.
      tx_size = 0
      consumer_sleep_=(0)
      ack_mode = "dups_ok"

      var consumer_clients = List[Client]()
      for (i <- 0 until destination_count) {
        val client = createConsumer(i)
        consumer_clients ::= client
        client.start()
      }

      // Keep sleeping until we stop draining messages.
      try {
        Thread.sleep(drain_timeout);
        def done() = {
          val c = consumer_counter.getAndSet(0)
          drained += c
          c == 0
        }
        while( !done ) {
          print(".")
          Thread.sleep(drain_timeout);
        }
      } finally {
        done.set(true)
        for( client <- consumer_clients ) {
          client.shutdown
        }
        for( client <- consumer_clients ) {
          client.wait_for_shutdown
        }
        println(". (drained %d)".format(drained))
        tx_size = original_tx_size
        _consumer_sleep = original_consumer_sleep
        ack_mode = original_ack_mode
      }
    }
  }


  def collection_sample: Unit = {

    val now = System.currentTimeMillis()
    val data = DataSample(
      time=now,
      produced=producer_counter.getAndSet(0),
      consumed=consumer_counter.getAndSet(0),
      errors=error_counter.getAndSet(0),
      max_latency=max_latency.getAndSet(0)
    )
    data_samples += data

    // we might need to increment number the producers..
    for (i <- 0 until producers_per_sample) {
      val client = createProducer(producer_clients.length)
      producer_clients ::= client
      client.start()
    }

    // we might need to increment number the consumers..
    for (i <- 0 until consumers_per_sample) {
      val client = createConsumer(consumer_clients.length)
      consumer_clients ::= client
      client.start()
    }

  }
  
  def createProducer(i:Int):Client
  def createConsumer(i:Int):Client

}

File: jms-benchmark/src/main/java/org/fusesource/jmsbenchmark/Scenario.java
/**
 * Copyright (C) 2009-2015 the original author or authors.
 * See the notice.md file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.fusesource.jmsbenchmark;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Supplier;
import static org.fusesource.jmsbenchmark.Support.*;

abstract public class Scenario {

    long NANOS_PER_SECOND = TimeUnit.NANOSECONDS.convert(1, TimeUnit.SECONDS);

    public static class Entry<A, B> extends AbstractMap.SimpleImmutableEntry<A, B> {
        public Entry(A key, B value) {
            super(key, value);
        }
    }

    String url = "tcp://localhost:61616";
    String userName;
    String password;

    SleepTask producerSleep = new SleepTask();
    void setProducerSleep(int value) {
        producerSleep = new FixedSleepTask(value);
    }

    SleepTask consumerSleep = new SleepTask();
    void setConsumerSleep(int value) {
        consumerSleep = new FixedSleepTask(value);
    }

    boolean allowWorkerInterrupt = false;
    boolean useMessageListener = false;

    int producers = 1;
    int producersPerSample = 0;

    int consumers = 1;
    int consumersPerSample = 0;
    int sampleInterval = 1000;

    int messageSize = 1024;
    boolean persistent = false;

    ArrayList<ArrayList<Entry<String, String>>> headers = new ArrayList<>();
    String selector = null;
    boolean noLocal = false;
    boolean durable = false;
    String ackMode = "auto";
    long messagesPerConnection = -1L;
    boolean displayErrors = false;
    int txSize = 0;

    String destinationType = "queue";
    Supplier<String> destinationName = () -> "load";

    String getDestinationName() {
        return destinationName.get();
    }
    void setDestinationName(String value) {
        destinationName = () -> value;
    }
    void setDestinationName(Supplier<String> value) {
        destinationName = value;
    }

    int destinationCount;

    AtomicLong producerCounter = new AtomicLong();
    AtomicLong consumerCounter = new AtomicLong();
    AtomicLong errorCounter = new AtomicLong();
    AtomicLong maxLatency = new AtomicLong();
    AtomicBoolean done = new AtomicBoolean();

    String queuePrefix = "";
    String topicPrefix = "";
    String name = "custom";

    long drainTimeout = 2000L;

    void run() throws TimeoutException, InterruptedException {
        System.out.print(toString());
        System.out.println("--------------------------------------");
        System.out.println("     Running: Press ENTER to stop");
        System.out.println("--------------------------------------");
        System.out.println("");

        withLoad(() -> {

            // start a sampling client...
            Thread sampleThread = new Thread() {

                void printRate(String name, long periodCount, long totalCount, long nanos) {
                    java.lang.Float ratePerSecond = ((1.0f * periodCount / nanos) * NANOS_PER_SECOND);
                    System.out.println("%s total: %,d, rate: %,.3f per second".format(name, totalCount, ratePerSecond));
                }

                public void run() {


                    try {
                        long start = System.nanoTime();
                        long totalProducerCount = 0L;
                        long totalConsumerCount = 0L;
                        long totalErrorCount = 0L;
                        collectionStart();
                        while (!done.get()) {
                            Thread.sleep(sampleInterval);
                            long end = System.nanoTime();
                            collectionSample();
                            ArrayList<DataSample> collected = collectionEnd();

                            if (producers > 0) {
                                long count = Support.last(collected).produced;
                                totalProducerCount += count;
                                if (totalProducerCount > 0) {
                                    printRate("Producer", count, totalProducerCount, end - start);
                                }
                            }

                            if (consumers > 0) {
                                long count = Support.last(collected).consumed;
                                totalConsumerCount += count;
                                if (totalConsumerCount > 0) {
                                    printRate("Consumer", count, totalConsumerCount, end - start);
                                }
                            }

                            long count = Support.last(collected).errors;
                            totalErrorCount += count;
                            if (totalErrorCount > 0) {
                                if (count != 0) {
                                    printRate("Error", count, totalErrorCount, end - start);
                                }
                            }

                            start = end;
                        }
                    } catch (InterruptedException e) {
                    }
                }
            };
            try {
                sampleThread.start();
                System.in.read();
                done.set(true);
                sampleThread.interrupt();
                sampleThread.join();
            } catch (Throwable e) {
                e.printStackTrace();
            }
            return null;
        });

    }

    @Override
    public String toString() {
        return "--------------------------------------\n" +
                "Scenario Settings\n" +
                "--------------------------------------\n" +
                "  destinationType       = " + destinationType + "\n" +
                "  queuePrefix           = " + queuePrefix + "\n" +
                "  topicPrefix           = " + topicPrefix + "\n" +
                "  destinationCount      = " + destinationCount + "\n" +
                "  destinationName       = " + getDestinationName() + "\n" +
                "  sampleInterval (ms)   = " + sampleInterval + "\n" +
                "  \n" +
                "  --- Producer Properties ---\n" +
                "  producers             = " + producers + "\n" +
                "  messageSize           = " + messageSize + "\n" +
                "  persistent            = " + persistent + "\n" +
                "  producerSleep (ms)    = " + producerSleep.get() + "\n" +
                "  headers               = " + mkString(headers, ", ") + "\n" +
                "  \n" +
                "  --- Consumer Properties ---\n" +
                "  consumers             = " + consumers + "\n" +
                "  consumerSleep (ms)    = " + consumerSleep.get() + "\n" +
                "  selector              = " + selector + "\n" +
                "  durable               = " + durable + "\n" +
                "";
    }

    protected ArrayList<Entry<String, String>> headersFor(int i) {
        if (headers.isEmpty()) {
            return new ArrayList<>();
        } else {
            return headers.get(i % headers.size());
        }
    }

    ArrayList<DataSample> dataSamples = new ArrayList<>();

    void collectionStart() {
        producerCounter.set(0);
        consumerCounter.set(0);
        errorCounter.set(0);
        maxLatency.set(0);
        dataSamples = new ArrayList<>();
    }

    void updateMaxLatency(long value) {
        long was = maxLatency.get();
        while (value > was && !maxLatency.compareAndSet(was, value)) {
            // someone else changed it before we could update it, lets
            // get the update and see if we still need to update.
            was = maxLatency.get();
        }
    }

    ArrayList<DataSample> collectionEnd() {
        return new ArrayList<>(dataSamples);
    }

    public static interface Client {
        int id();

        void start();

        void shutdown();

        void waitForShutdown() throws InterruptedException;
    }

    ArrayList<Client> producerClients = new ArrayList<>();
    ArrayList<Client> consumerClients = new ArrayList<>();

    CountDownLatch connectedLatch = null;
    CountDownLatch applyLoadLatch = null;

    <T> T withLoad(Supplier<T> func) throws TimeoutException, InterruptedException {
        final int[] i = new int[]{0};
        return withLoadAndConnectTimeout(func, () -> {
            i[0] += 1;
            if (i[0] <= 10) {
                return 1000L;
            } else {
                return 0L;
            }
        });
    }

    <T> T withLoadAndConnectTimeout(Supplier<T> func, Supplier<Long> continueConnecting) throws TimeoutException, InterruptedException {
        done.set(false);

        producerSleep.init(System.currentTimeMillis());
        consumerSleep.init(System.currentTimeMillis());

        connectedLatch = new CountDownLatch(producers + consumers);
        applyLoadLatch = new CountDownLatch(1);

        for (int i = 0; i < producers; i++) {
            Client client = createProducer(i);
            producerClients.add(client);
            client.start();
        }

        for (int i = 0; i < consumers; i++) {
            Client client = createConsumer(i, false);
            consumerClients.add(client);
            client.start();
        }

        try {
            long timeout = 0L;
            boolean doneConnect = false;
            while (!doneConnect) {
                if (connectedLatch.await(timeout, TimeUnit.MILLISECONDS)) {
                    doneConnect = true;
                    applyLoadLatch.countDown();
                } else {
                    timeout = continueConnecting.get();
                    if (timeout <= 0) {
                        applyLoadLatch.countDown();
                        done.set(true);
                        throw new TimeoutException("All clients could not connect.  Failing connections: " + connectedLatch.getCount());
                    }
                }
            }
            return func.get();

        } finally {
            done.set(true);
            for (Client client : producerClients) {
                client.shutdown();
            }
            for (Client client : producerClients) {
                client.waitForShutdown();
            }
            producerClients = new ArrayList<>();
            // wait for the threads to finish..
            for (Client client : consumerClients) {
                client.shutdown();
            }
            for (Client client : consumerClients) {
                client.waitForShutdown();
            }
            consumerClients = new ArrayList<>();
        }
    }

    void drain() throws InterruptedException {
        done.set(false);
        if ("queue".equals(destinationType) || "raw_queue".equals(destinationType) || durable == true) {
            System.out.print("draining");
            consumerCounter.set(0);
            final long[] drained = new long[]{ 0L };

            int originalTxSize = txSize;
            SleepTask originalConsumerSleep = consumerSleep;
            String originalAckMode = ackMode;

            // Lets change the consumer config to consume as fast as possible.
            txSize = 0;
            setConsumerSleep(0);
            ackMode = "dups_ok";

            ArrayList<Client> consumerClients = new ArrayList<Client>();

            int drainingConsumers = "queue".equals(destinationType) ? destinationCount : consumers;
            for (int i = 0; i < drainingConsumers; i++) {
                Client client = createConsumer(i, true);
                consumerClients.add(client);
                client.start();
            }

            // Keep sleeping until we stop draining messages.
            try {
                Thread.sleep(drainTimeout);
                Supplier<Boolean> done = () -> {
                        long c = consumerCounter.getAndSet(0);
                        drained[0] += c;
                        return c == 0;
                };
                while (!done.get()) {
                    System.out.print(".");
                    Thread.sleep(drainTimeout);
                }
            } finally {
                done.set(true);
                for (Client client : consumerClients) {
                    client.shutdown();
                }
                for (Client client : consumerClients) {
                    client.waitForShutdown();
                }
                System.out.printf(". (drained %d)", drained[0]);
                txSize = originalTxSize;
                consumerSleep = originalConsumerSleep;
                ackMode = originalAckMode;
            }
        }
    }


    void collectionSample() {

        long now = System.currentTimeMillis();
        DataSample data = new DataSample(now,
                producerCounter.getAndSet(0),
                consumerCounter.getAndSet(0),
                errorCounter.getAndSet(0),
                maxLatency.getAndSet(0)
        );
        dataSamples.add(data);

        // we might need to increment number the producers..
        for (int i = 0; i < producersPerSample; i++) {
            Client client = createProducer(producerClients.size());
            producerClients.add(client);
            client.start();
        }

        // we might need to increment number the consumers..
        for (int i = 0; i < consumersPerSample; i++) {
            Client client = createConsumer(consumerClients.size(), false);
            consumerClients.add(client);
            client.start();
        }

    }

    abstract Client createProducer(int i);

    abstract Client createConsumer(int i, boolean drain);

}

